### PR TITLE
Feat/open file dialog

### PR DESCRIPTION
**Pull request by justinvdk**
File dialog after feedback. With this implementation one can select a directory in which the files can be found.

Maven doesn't pass for me, sometimes a test just hangs, not sure where that is coming from.
Other tests fail because this feature makes it so that the application needs files as input, most (if not all) hardcodes references in non-test code is gone, and thus no input data is loaded automatically. Some tests expect the application to start, having loaded at least some data (TB10.gfa usually).

For now I can't come up with a clean way to give the tests an opportunity to do their thing without compromising code-cleanness.

### CODE REVIEW

**Comment by wouterjsmit** [comment]
> Some tests expect the application to start, having loaded at least some data (TB10.gfa usually).

Would have been avoided with good practice unit testing.

**Comment by FarisElghlan** [comment]
The tests which use that file are integration tests. It was not possible to select a different file with the old code (well I could probably change the file name to a variable and change it during tests, but then the "real" application isn't tested).
May I fix the integration test or are you still working on something else which will cause merge conflicts?

**Comment by justinvdk** [comment]
~~@FarisElghlan I dont have any more features to implement so I don't think fixing the integration tests would hit any further changes from my side.~~

**Comment by justinvdk** [comment]
@FarisElghlan I will first see if I can decouple the file->model object logic. This might cause merge conflics. I will update this thread when I got more info.

**Comment by justinvdk** [comment]
There are indeed conlicts now. Mainly in RootLayoutcontroller.
I've attempted to merge once, code looked nice but when I manually tested, nothing happened when dragging+dropping nodes.
Any ideas?

**Comment by FarisElghlan** [comment]
Did you use the TB10 graph and tree? the 328 graph / 340 tree don't have the same genome names yet (fixed like 5 minutes ago in a branch which wouter is still working on, so might take some time till it will be merged to dev, but will be fixed relatively soon)

**Comment by justinvdk** [comment]
Would that result in literally zero graph nodes showing up in the comparison? If so, I'll try again with TB10 graph and tree.

**Comment by FarisElghlan** [comment]
Yes, because the names don't match, the graph acts as if you drop a couple of strings in there which don't correspond to any of the genomes in the graph, so nothing happens.

**Comment by justinvdk** [comment]
That indeed seems to be the problem. However, I get an exception when dragging+dropping now (when dropping). I did not put a lot of effort into debugging because I have no clue where to look. I'll include a stacktrace for reference but will look further tomorrow.

```
WARNING: null
Exception in thread "JavaFX Application Thread" java.lang.ArrayIndexOutOfBoundsException: -1
    at java.util.ArrayList.elementData(ArrayList.java:418)
    at java.util.ArrayList.get(ArrayList.java:431)
    at nl.tudelft.pl2016gr2.core.algorithms.subgraph.CompareSubgraphs.removeEmptyLevels(CompareSubgraphs.java:62)
    at nl.tudelft.pl2016gr2.core.algorithms.subgraph.SubgraphAlgorithmManager.alignOneGraph(SubgraphAlgorithmManager.java:75)
    at nl.tudelft.pl2016gr2.gui.view.graph.DrawComparedGraphs.drawOneGraph(DrawComparedGraphs.java:382)
    at nl.tudelft.pl2016gr2.gui.view.graph.DrawComparedGraphs.redrawGraphs(DrawComparedGraphs.java:367)
    at nl.tudelft.pl2016gr2.gui.view.graph.DrawComparedGraphs.handleGenomesDropped(DrawComparedGraphs.java:251)
    at nl.tudelft.pl2016gr2.gui.view.graph.DrawComparedGraphs.lambda$initializeTopDragEvent$4(DrawComparedGraphs.java:205)
    at nl.tudelft.pl2016gr2.gui.view.graph.DrawComparedGraphs$$Lambda$81/930260377.handle(Unknown Source)
    at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
    at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
    at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
    at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
    at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
    at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
(...) snipped a whole bunch of javafx/sun stacktraces
```

**Comment by FarisElghlan** [comment]
Hmm that basically means the graph contained no nodes after it was split. I guess something (ex. the maingraph or maingraphorder in drawcompared graphs) is not initialized correctly. I currently at work, will have time to look at it this evening.

**Comment by justinvdk** [comment]
Alright. I can't seem to find it to be honest. 
When leave everything in place and replace line 66 and 72 in [RootLayoutController](https://github.com/ProgrammingLife2016/PL2-2016/blob/2486679f7c68475ed1efb55c55c34ddc732c6fb6/PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/RootLayoutController.java##L62) with loading the data, all goes well. So something goes wrong when not immediately loading the graph+tree.

**Comment by FarisElghlan** [comment]
When loading the TB10 file everything works for me (the TB328 file got a ".fasta" extension, so I guess it didn't work because your branch couldn't handle these yet).
I am currently fixing the tests. The file browser works really nice by the way!

File: PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/RootLayoutController.java [hunk 1]
```diff
@@ -162,4 +162,13 @@ public void filesLoaded(InputStream treeFile, InputStream graphFile, InputStream
       Logger.getLogger(getClass().getName()).log(Level.SEVERE, null, e);
     }
   }
+
+  @FXML
+  @SuppressWarnings("unused")
+  private void openFileMenuItemClicked() throws IOException {
```
Comment (justinvdk) [inline PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/RootLayoutController.java]: Maybe we can call this method from [here](https://github.com/ProgrammingLife2016/PL2-2016/blob/feat/open-file-dialog/PL2/PL2-launcher/src/main/java/nl/tudelft/pl2016gr2/launcher/Dnav.java#L45)?

With this the SupressWarnings can go too.

File: PL2/PL2-core/pom.xml [hunk 2]
```diff
@@ -31,6 +31,18 @@
             <version>1.0-SNAPSHOT</version>
         </dependency>
         <dependency>
+            <groupId>org.mockito</groupId>
```
Comment (justinvdk) [inline PL2/PL2-core/pom.xml]: Mockito is already defined a bit lower.

edit: Which is a change I did apparently :) I think this snuck into the branch with a merge. I'll take care of this before rebasing.

File: PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/FileGraphFactory.java [hunk 3]
```diff
@@ -4,14 +4,14 @@
 import nl.tudelft.pl2016gr2.model.SequenceGraph;
 import nl.tudelft.pl2016gr2.parser.controller.GfaReader;
 
-import java.io.File;
+import java.io.InputStream;
 
 public class FileGraphFactory implements GraphFactory {
```
Comment (justinvdk) [inline PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/FileGraphFactory.java]: Maybe rename the class to StreamGraphFactory

File: PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/FileTreeFactory.java [hunk 4]
```diff
@@ -4,30 +4,26 @@
...
-   * @param file The file to read
+   * @param fileStream The file to read
    * @throws FileNotFoundException when file does not exist
    */
-  public FileTreeFactory(File file) throws FileNotFoundException {
-    this.file = file;
-    if (!file.exists()) {
-      throw new FileNotFoundException();
-    }
+  public FileTreeFactory(InputStream fileStream) throws FileNotFoundException {
```
Comment (justinvdk) [inline PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/FileTreeFactory.java]: Maybe rename to StreamTreeFactory

File: PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/RootLayoutController.java [hunk 5]
```diff
@@ -137,4 +138,37 @@ private void initializeSelectionManager() {
...
+            Level.SEVERE,
+            "tree or graph was null");
+      }
+    } catch (IOException e) {
+      Logger.getLogger(getClass().getName()).log(Level.SEVERE, null, e);
+    }
+  }
+
+  @FXML
+  @SuppressWarnings("unused")
```
Comment (justinvdk) [inline PL2/PL2-gui/src/main/java/nl/tudelft/pl2016gr2/gui/view/RootLayoutController.java:181]: Maybe we can call this method from here?

With this the SupressWarnings can go too.

**Pull request merged**